File: crypto/pqc.py
"""
crypto/pqc.py
==============
Post-quantum cryptography via liboqs-python 0.16.0.

Algorithms used
---------------
- ML-KEM-768  (NIST FIPS 203, formerly Kyber-768) for key encapsulation.
- ML-DSA-65   (NIST FIPS 204, formerly Dilithium3) for signing.

Both names are the standardized NIST identifiers used since liboqs 0.11.0.
The old Kyber*/Dilithium* aliases were removed in liboqs 0.11.x.

Public API
----------
KEM
  kem_generate_keypair()      -> (public_key_bytes, secret_key_bytes)
  kem_encapsulate(pk_bytes)   -> (ciphertext_bytes, shared_secret_bytes)
  kem_decapsulate(sk, ct)     -> shared_secret_bytes

Signing
  sign_generate_keypair()         -> (public_key_bytes, secret_key_bytes)
  sign_message(sk, message)       -> signature_bytes
  verify_signature(pk, message, signature) -> None  (raises on failure)

Exceptions
----------
SignatureVerificationError
    Raised by verify_signature() when verification fails.
    The exception message contains the specific reason string.
"""

from __future__ import annotations

import oqs

# ── Algorithm identifiers ────────────────────────────────────────────────────
_KEM_ALG = "ML-KEM-768"
_SIG_ALG = "ML-DSA-65"


# ── Custom exceptions ────────────────────────────────────────────────────────

class SignatureVerificationError(Exception):
    """Raised when ML-DSA signature verification fails.

    The message attribute contains the specific structured reason string
    displayed in Eve's attack-result console.
    """


# ── KEM: ML-KEM-768 ─────────────────────────────────────────────────────────

def kem_generate_keypair() -> tuple[bytes, bytes]:
    """Generate an ML-KEM-768 keypair.

    Returns
    -------
    public_key_bytes : bytes
        1184-byte public key. Share with the encapsulating party.
    secret_key_bytes : bytes
        2400-byte secret key. Keep private.
    """
    with oqs.KeyEncapsulation(_KEM_ALG) as kem:
        public_key_bytes = kem.generate_keypair()
        secret_key_bytes = kem.export_secret_key()
    return public_key_bytes, secret_key_bytes


def kem_encapsulate(peer_public_key_bytes: bytes) -> tuple[bytes, bytes]:
    """Encapsulate a shared secret for the peer's ML-KEM-768 public key.

    Parameters
    ----------
    peer_public_key_bytes : bytes
        The peer's ML-KEM-768 public key.

    Returns
    -------
    ciphertext_bytes : bytes
        1088-byte ciphertext to send to the peer.
    shared_secret_bytes : bytes
        32-byte shared secret. Do not transmit.
    """
    with oqs.KeyEncapsulation(_KEM_ALG) as kem:
        ciphertext_bytes, shared_secret_bytes = kem.encap_secret(
            peer_public_key_bytes
        )
    return ciphertext_bytes, shared_secret_bytes


def kem_decapsulate(secret_key_bytes: bytes, ciphertext_bytes: bytes) -> bytes:
    """Decapsulate to recover the shared secret.

    Parameters
    ----------
    secret_key_bytes : bytes
        Our ML-KEM-768 secret key.
    ciphertext_bytes : bytes
        Ciphertext received from the encapsulating party.

    Returns
    -------
    bytes
        32-byte shared secret. Must equal the encapsulator's shared_secret.
    """
    with oqs.KeyEncapsulation(_KEM_ALG, secret_key=secret_key_bytes) as kem:
        shared_secret_bytes = kem.decap_secret(ciphertext_bytes)
    return shared_secret_bytes


# ── Signatures: ML-DSA-65 ────────────────────────────────────────────────────

def sign_generate_keypair() -> tuple[bytes, bytes]:
    """Generate an ML-DSA-65 signing keypair.

    Returns
    -------
    public_key_bytes : bytes
        1952-byte public key. Distribute to verifiers.
    secret_key_bytes : bytes
        4032-byte secret key. Keep private.
    """
    with oqs.Signature(_SIG_ALG) as signer:
        public_key_bytes = signer.generate_keypair()
        secret_key_bytes = signer.export_secret_key()
    return public_key_bytes, secret_key_bytes


def sign_message(secret_key_bytes: bytes, message: bytes) -> bytes:
    """Sign an arbitrary message with ML-DSA-65.

    Parameters
    ----------
    secret_key_bytes : bytes
        Our ML-DSA-65 secret key.
    message : bytes
        The message to sign (typically a handshake transcript hash).

    Returns
    -------
    bytes
        ML-DSA-65 signature (variable length, ≤ 3309 bytes for ML-DSA-65).
    """
    with oqs.Signature(_SIG_ALG, secret_key=secret_key_bytes) as signer:
        signature = signer.sign(message)
    return signature


def verify_signature(
    public_key_bytes: bytes,
    message: bytes,
    signature: bytes,
) -> None:
    """Verify an ML-DSA-65 signature.

    Raises
    ------
    SignatureVerificationError
        If verification fails. The exception message is the structured reason
        string shown to Eve in the attack-result console, e.g.:
            "ML-DSA signature verification failed: transcript hash mismatch"

    Returns
    -------
    None
        Returns normally (implicitly None) on success.
    """
    with oqs.Signature(_SIG_ALG) as verifier:
        valid = verifier.verify(message, signature, public_key_bytes)
    if not valid:
        raise SignatureVerificationError(
            "ML-DSA signature verification failed: transcript hash mismatch"
        )

File: frontend/src/crypto/pqc.js
/**
 * pqc.js
 * ML-KEM-768 (FIPS 203) and ML-DSA-65 (FIPS 204) via @noble/post-quantum 0.7.x.
 *
 * Mirrors crypto/pqc.py exactly in terms of what each function returns and throws.
 * Same exception vocabulary:
 *   SignatureVerificationError: "ML-DSA signature verification failed: transcript hash mismatch"
 *   verifySignature() throws on failure, returns undefined on success.
 *
 * @noble/post-quantum 0.7.x API (differs from naive expectations — read carefully):
 *   keygen()             not  generateKeyPair()
 *   result.secretKey     not  result.privateKey
 *   encapsulate()        returns { cipherText, sharedSecret }  (capital T in cipherText)
 *   decapsulate(ct, sk)  cipherText first, secretKey second
 *   sign(msg, sk)        message is FIRST argument, secretKey is SECOND
 *   verify(sig, msg, pk) signature is FIRST, message SECOND, publicKey THIRD
 */

import { ml_kem768 } from '@noble/post-quantum/ml-kem.js';
import { ml_dsa65  } from '@noble/post-quantum/ml-dsa.js';

// ── Custom exception (exact string used throughout the project) ────────────

export class SignatureVerificationError extends Error {
  constructor(reason) {
    super(reason);
    this.name = 'SignatureVerificationError';
  }
}

// ── ML-KEM-768 ─────────────────────────────────────────────────────────────

/**
 * Generate an ML-KEM-768 keypair.
 * @returns {{ publicKey: Uint8Array, secretKey: Uint8Array }}
 *   publicKey = 1184 bytes, secretKey = 2400 bytes
 */
export function kemGenerateKeypair() {
  return ml_kem768.keygen();
  // Returns { publicKey, secretKey }
}

/**
 * Encapsulate to a public key (Bob encapsulates to Alice's kemPk).
 * @param {Uint8Array} publicKey  1184-byte ML-KEM-768 public key
 * @returns {{ cipherText: Uint8Array, sharedSecret: Uint8Array }}
 *   cipherText = 1088 bytes, sharedSecret = 32 bytes
 */
export function kemEncapsulate(publicKey) {
  return ml_kem768.encapsulate(publicKey);
  // Returns { cipherText, sharedSecret }
}

/**
 * Decapsulate to recover the shared secret (Alice decapsulates with her secretKey).
 * Note: decapsulate NEVER throws — ML-KEM returns a random value on wrong key.
 * Wrong session key will be caught later by GCM authentication failure.
 *
 * @param {Uint8Array} secretKey  2400-byte ML-KEM-768 secret key
 * @param {Uint8Array} cipherText 1088-byte ciphertext from encapsulate()
 * @returns {Uint8Array} 32-byte shared secret
 */
export function kemDecapsulate(secretKey, cipherText) {
  return ml_kem768.decapsulate(cipherText, secretKey);
  // Note argument order: (cipherText, secretKey) in noble API
}

// ── ML-DSA-65 ──────────────────────────────────────────────────────────────

/**
 * Generate an ML-DSA-65 keypair.
 * @returns {{ publicKey: Uint8Array, secretKey: Uint8Array }}
 *   publicKey = 1952 bytes, secretKey = 4032 bytes
 */
export function signGenerateKeypair() {
  return ml_dsa65.keygen();
  // Returns { publicKey, secretKey }
}

/**
 * Sign a message with ML-DSA-65.
 * @param {Uint8Array} secretKey  4032-byte ML-DSA-65 secret key
 * @param {Uint8Array} message    Bytes to sign (the 32-byte transcript hash)
 * @returns {Uint8Array} Signature (up to 3309 bytes)
 */
export function signMessage(secretKey, message) {
  // noble API: sign(message, secretKey) — message is FIRST
  return ml_dsa65.sign(message, secretKey);
}

/**
 * Verify a ML-DSA-65 signature over a message.
 * Throws on failure. Returns undefined on success (matching Python contract).
 *
 * @param {Uint8Array} publicKey  1952-byte ML-DSA-65 public key
 * @param {Uint8Array} message    The signed bytes (transcript hash)
 * @param {Uint8Array} signature  The signature to verify
 * @throws {SignatureVerificationError}
 *   "ML-DSA signature verification failed: transcript hash mismatch"
 */
export function verifySignature(publicKey, message, signature) {
  // noble API: verify(signature, message, publicKey) — sig is FIRST
  const valid = ml_dsa65.verify(signature, message, publicKey);
  if (!valid) {
    throw new SignatureVerificationError(
      'ML-DSA signature verification failed: transcript hash mismatch'
    );
  }
  // Returns undefined on success — same contract as Python reference
}
